lvsfunc/comparison.py

```diff
@@ -27,111 +27,64 @@ class Direction(IntEnum):
-        self.height: Optional[int] = self.heights.pop() if len(self.heights) == 1 else None
-
-        if self.width == 0:
-            self.width = None
-        if self.height == 0:
-            self.height = None  # this should be fixed on VapourSynth's side since variable-format uses None but width/height uses 0
-
-        self.formats: List[Union[None, str]] = []
-        for clip, name in self.clips:
-            try:
-                self.formats.append(clip.format.name)
-            except AttributeError:
-                self.formats.append(clip.format)
-        self.format: Optional[str] = self.formats[0] if len(set(self.formats)) == 1 else None
-
-    def _smart_crop(self, direction: Direction = Direction.HORIZONTAL):
-        if not self.width or not self.height:
-            raise ValueError("Comparer: all clips must have same width and height if cropping")
-        breaks_subsampling = ((direction == Direction.HORIZONTAL and (((self.width // self.num_clips) % 2)
-                                                                      or ((self.width % self.num_clips) % 2)))
-                              or (direction == Direction.VERTICAL and (((self.height // self.num_clips) % 2)
-                                                                       or ((self.height % self.num_clips) % 2))))
-
-        if breaks_subsampling and not all(get_subsampling(clip) in ('444', None) for clip in self.clips.values()):
-            raise ValueError("Comparer: resulting cropped width or height violates subsampling rules; "
-                             "consider resampling to YUV444 or RGB before attempting to crop")
-
-        if direction == Direction.HORIZONTAL:
-            crop_width, overflow = self.width // self.num_clips, self.width % self.num_clips
-
-            for key, (name, clip) in enumerate(self.clips.items()):
-                left_crop = crop_width * key
-                right_crop = crop_width * (self.num_clips - 1 - key)
-
-                if key != (self.num_clips - 1):
-                    right_crop += overflow
+        self.clips: List[vs.VideoNode] = list(clips.values()) if isinstance(clips, dict) else list(clips)
+        self.names: Optional[List[str]] = list(clips.keys()) if isinstance(clips, dict) else None
 
-                self.clips[name] = clip.std.Crop(left=left_crop, right=right_crop)
+        self.label_alignment = label_alignment
 
-        elif direction == Direction.VERTICAL:
-            crop_height, overflow = self.height // self.num_clips, self.height % self.num_clips
+        self.num_clips = len(clips)
 
-            for key, (name, clip) in enumerate(self.clips.items()):
-                top_crop = crop_height * key
-                bottom_crop = crop_height * (self.num_clips - 1 - key)
+        widths: Set[int] = {clip.width for clip in self.clips}
+        self.width: Optional[int] = w if (w := widths.pop()) != 0 and len(widths) == 0 else None
```

Comment:
You need to check `len(widths)` first and that should be `== 1`. Same for the following two attributes.

Reply:
For `format` yes, that was a typo, but for these I'm pretty sure this is correct even if messy. The assignment operator (the pop) happens first, so the length should become `0` if there's only one width in the set. Tested with Stack and this seems to work how I intended. If there's a cleaner way of doing this on one line I'd change it since I didn't really give it much thought.

Reply:
> The assignment operator (the pop) happens first, so the length should become `0` if there's only one width in the set.

Ah, my thought was that pop could fail if `width` is empty, but that could only happen if `self.clips` was empty, which is bad usage in the first place. Regardless, I believe checking the length first and *then* checking the value of the only entry is more true to what happens.

Reply:
I have a check to make sure len(clips) >= 2 so self.clips can't be empty.

Reply:
The way it works currently is to prevent a "w might be referenced before assignment" error because checking the list length first and then popping *might not* pop the value if the length bool is false.

Reply:
Sure, but then the expression wouldn't evaluate to true and the `if-true` expression isn't evaluated. In reality, it doesn't make a difference here, but it would if `widths` was empty.

Just so it's clear what I'm talking about:
```suggestion
        self.width: Optional[int] = w if len(widths) == 1 and (w := widths.pop()) != 0 else None
```